dcos/nodeutil/util.go

```diff
@@ -245,6 +246,20 @@ func (d *dcosInfo) MesosID() (string, error) {
 		panic("Unable to get mesos ID. Uninitialized url")
 	}
 
+	role, err := d.Role()
+	if err != nil {
+		return "", err
+	}
+
+	// if we request IP for an agent node, we should look at `slaves` field.
+	localIP, err := d.DetectIP()
+	if err != nil {
+		return "", err
+	}
+
+	d.Lock()
```

Comment:
why are you locking around blocking i/o code? seems to me that you can make the HTTP request and also process the response status without the locking.